### PR TITLE
feat: parse EXT-X-DATERANGE tags and attributes

### PR DESCRIPTION
**Pull request by adrums86**
Add support for `EXT-X-DATERANGE` tag parsing to the HLS parser.

https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis#section-4.4.5.1

### CODE REVIEW

File: packages/web-media-box/src/hls-parser/consts/tags.ts [hunk 1]
```diff
@@ -13,11 +13,15 @@ export const EXT_X_PART_INF = 'EXT-X-PART-INF';
...
-export const EXT_X_MAP = 'EXT_X_MAP';
+export const EXT_X_DISCONTINUITY = 'EXT-X-DISCONTINUITY';
+export const EXT_X_KEY = 'EXT-X-KEY';
+export const EXT_X_MAP = 'EXT-X-MAP';
 export const EXT_X_GAP = 'EXT-X-GAP';
 export const EXT_X_PROGRAM_DATE_TIME = 'EXT-X-PROGRAM-DATE-TIME';
 export const EXT_X_BITRATE = 'EXT-X-BITRATE';
 export const EXT_X_PART = 'EXT-X-PART';
-export const EXT_X_MEDIA = 'EXT_X_MEDIA';
+export const EXT_X_MEDIA = 'EXT-X-MEDIA';
```
Comment (adrums86) [inline packages/web-media-box/src/hls-parser/consts/tags.ts:23]: Just fixing a couple of sneaky underscores here 😄

**Review by wseymour15 — APPROVED** [review]
Makes sense to me, nice work!

File: packages/web-media-box/src/hls-parser/tags/tagWithAttributesProcessors.ts [hunk 2]
```diff
@@ -240,3 +240,54 @@ export class ExtXMedia extends TagWithAttributesProcessor {
...
+    // If we have more attributes on the tag than already parsed we have client attributes.
+    if (Object.keys(dateRange).length < attributeKeys.length) {
+      dateRange.clientAttributes = {};
+      attributeKeys.forEach((key) => {
+        if (key.startsWith(ExtXDaterange.CLIENT_ATTRIBUTES) && dateRange.clientAttributes) {
+          // The attribute value MUST be a quoted-string, a hexadecimal-sequence, or a decimal-floating-point.
+          dateRange.clientAttributes[key] = Number.isNaN(Number(tagAttributes[key])) ? tagAttributes[key] : Number(tagAttributes[key]);
+        }
+      });
+    }
```
Comment (dzianis-dashkevich) [inline packages/web-media-box/src/hls-parser/tags/tagWithAttributesProcessors.ts]: we can probably omit the whole check and do the following:

```ts
Object
.keys(tagAttributes)
.filter((tagKey) => tagKey.startsWith(ExtXDaterange.CLIENT_ATTRIBUTES))
.reduce((clientAttributes, tagKey) => { ... }, dateRange.clientAttributes)

```

we should set `clientAttributes` by default to an empty object

File: packages/web-media-box/src/hls-parser/types/parsedPlaylist.d.ts [hunk 3]
```diff
@@ -81,6 +81,27 @@ export interface RenditionGroups {
...
+
+export interface DateRange {
+  id: string;
+  class?: string;
+  startDate: string;
+  cue?: Cue[];
+  endDate?: string;
+  duration?: number;
+  plannedDuration?: number;
+  clientAttributes?: Record<string, string | number>;
```
Comment (dzianis-dashkevich) [inline packages/web-media-box/src/hls-parser/types/parsedPlaylist.d.ts]: can be not optional

File: packages/web-media-box/src/hls-parser/types/parsedPlaylist.d.ts [hunk 4]
```diff
@@ -81,6 +81,27 @@ export interface RenditionGroups {
...
+  scte35Cmd?: number;
+  scte35Out?: number;
+  scte35In?: number;
+  endOnNext?: boolean;
```
Comment (dzianis-dashkevich) [inline packages/web-media-box/src/hls-parser/types/parsedPlaylist.d.ts]: can be not optional

File: packages/web-media-box/src/hls-parser/types/parsedPlaylist.d.ts [hunk 5]
```diff
@@ -117,4 +138,6 @@ export interface ParsedPlaylist {
   renditionGroups: RenditionGroups;
   // Used to persist EXT_X_BITRATE across segments
   currentBitrate?: number;
+  // https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis#section-4.4.5.1
+  dateRanges?: Array<DateRange>;
```
Comment (dzianis-dashkevich) [inline packages/web-media-box/src/hls-parser/types/parsedPlaylist.d.ts]: Can be empty array by default

**Review by dzianis-dashkevich — APPROVED** [review]

File: packages/web-media-box/src/hls-parser/tags/tagWithAttributesProcessors.ts [hunk 6]
```diff
@@ -240,3 +240,51 @@ export class ExtXMedia extends TagWithAttributesProcessor {
...
+  private static readonly SCTE35_IN = 'SCTE35-IN';
+  private static readonly END_ON_NEXT = 'END-ON-NEXT';
+
+  protected requiredAttributes = new Set([ExtXDaterange.ID, ExtXDaterange.START_DATE]);
+  protected tag = EXT_X_DATERANGE;
+
+  protected safeProcess(tagAttributes: Record<string, string>, playlist: ParsedPlaylist): void {
+    if (!playlist.dateRanges) {
+      playlist.dateRanges = [];
+    }
```
Comment (dzianis-dashkevich) [inline packages/web-media-box/src/hls-parser/tags/tagWithAttributesProcessors.ts]: you can move it to the playlist initializer and you can probably remove `?`  from `dateRanges?: DateRange[];` model

**Review by dzianis-dashkevich — APPROVED** [review]

File: packages/web-media-box/src/hls-parser/tags/tagWithAttributesProcessors.ts (hunk 6 above)
Comment (adrums86) [reply packages/web-media-box/src/hls-parser/tags/tagWithAttributesProcessors.ts]: Good call, fixed.

**Pull request merged**